tasktiger/lua/move_task.lua

```diff
@@ -0,0 +1,85 @@
+local function zadd_w_mode(key, score, member, mode)
+    if mode == "" then
+        redis.call('zadd', key, score, member)
+    elseif mode == "nx" then
+        zadd_noupdate({ key }, { score, member })
+    elseif mode == "xx" then
+        zadd_update_existing({ key }, { score, member })
+    elseif mode == "min" then
+        zadd_update_min({ key }, { score, member })
+    elseif mode == "max" then
+        zadd_update_max({ key }, { score, member })
+    else
+        error("mode " .. mode .. " unsupported")
+    end
+end
+
+
+local function get_key_func(key_prefix)
+    return function(...)
+        table.insert(arg, 1, key_prefix)
+        return table.concat(arg, ':')
+    end
+end
+
+local key_prefix = ARGV[1]
+local id = ARGV[2]
+local queue = ARGV[3]
+local from_state = ARGV[4]
+local to_state = ARGV[5]
+local unique = ARGV[6]
+local when = ARGV[7]
+local mode = ARGV[8]
+local publish_queued_tasks = ARGV[9]
+
+local key = get_key_func(key_prefix)
+
+assert(redis.call('zscore', key(from_state, queue), id), '<FAIL_IF_NOT_IN_ZSET>')
+
+if to_state ~= "" then
+    zadd_w_mode(key(to_state, queue), when, id, mode)
+    redis.call('sadd', key(to_state), queue)
+end
+redis.call('zrem', key(from_state, queue), id)
+
+if to_state == "" then -- Remove the task if necessary
+    if unique == 'true' then
+        -- Delete executions if there were no errors
+        local to_delete = {
+            key('task', id, 'executions'),
```

Comment:
Since we're rewriting this from scratch, we shouldn't be generating keys inside lua scripts.

From [Redis docs](https://redis.io/docs/manual/programmability/eval-intro/):
```
At this point, it is essential to understand the distinction Redis makes between input arguments that are names of keys and those that aren't.

While key names in Redis are just strings, unlike any other string values, these represent keys in the database. The name of a key is a fundamental concept in Redis and is the basis for operating the Redis Cluster.

Important: to ensure the correct execution of scripts, both in standalone and clustered deployments, all names of keys that a script accesses must be explicitly provided as input key arguments. The script should only access keys whose names are given as input arguments. Scripts should never access keys with programmatically-generated names or based on the contents of data structures stored in the database.
```

Reply:
done in latest commit!